aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+        // Extract columns:
+        vol2col<scalar_t>(
+            input_n.data<scalar_t>(),
+            nInputPlane,
+            inputDepth,
+            inputHeight,
+            inputWidth,
+            outputDepth,
+            outputHeight,
+            outputWidth,
+            kD,
+            kH,
+            kW,
+            padD,
+            padH,
+            padW,
+            dD,
+            dH,
+            dW,
+            dilationD,
+            dilationH,
+            dilationW,
+            columns.data<scalar_t>());
+        // For gemm argument derivation, see
+        // conv_dilated2d_all_cuda_template in
+        // ATen/native/cuda/DilatedConvolution.cu
+        THBlas_gemm<scalar_t>(
+            /*transa=*/'t',
+            /*transb=*/'n',
+            /*     m=*/columns.size(0),
+            /*     n=*/nOutputPlane,
+            /*     k=*/columns.size(1),
+            /* alpha=*/scale,
+            /*     A=*/columns.data<scalar_t>(),
+            /*   lda=*/columns.size(1),
+            /*     B=*/grad_output_n.data<scalar_t>(),
+            /*   ldb=*/columns.size(1),
+            /*  beta=*/1,
+            /*     C=*/grad_weight.data<scalar_t>(),
+            /*   ldc=*/columns.size(0));
+      }
+
+      // Gradient of bias:
+      if (grad_bias.defined()) {
+        grad_bias += grad_output_n.sum({1, 2, 3});
+      }
+    }
+  });
+  if (is_batch) {
+    DROP_BATCH_DIMENSION(input);
+    DROP_BATCH_DIMENSION(output);
+    DROP_BATCH_DIMENSION(grad_input);
+    DROP_BATCH_DIMENSION(grad_output);
+  }
+
+} // conv_dilated3d_all_cpu_template
+
+} // namespace
+
+Tensor conv_dilated2d_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    const Tensor& bias,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size) {
+  auto options = input.options();
+  Tensor output = at::empty({0}, options);
+  const Tensor weight_ = weight.contiguous();
+  const Tensor input_ = input.contiguous();
+  const Tensor bias_ = bias.contiguous();
+  Tensor undefined;
+  int64_t nOutputPlane = weight.size(0);
+  int64_t outputHeight = OUTPUTSIZE(0, 2);
+  int64_t outputWidth = OUTPUTSIZE(1, 2);
+  output.resize_({nOutputPlane, outputHeight, outputWidth});
+  if (input.dim() == 4) {
+    INSERT_BATCH_DIMENSION(output, input.size(0));
+  }
+  conv_dilated2d_all_cpu_template(
+      output,
+      input_,
+      weight_,
+      bias_,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      kernel_size,
+      stride_size,
+      pad_size,
+      dilation_size);
+  return output;
+}
+
+std::tuple<Tensor, Tensor, Tensor> conv_dilated2d_backward_cpu(
+    const Tensor& grad_output,
+    const Tensor& input,
+    const Tensor& weight,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size,
+    const std::array<bool, 3ul> output_mask) {
```

Comment:
You are ignoring `output_mask` here, which to me indicates immediately something has gone wrong. Let me explain.

When we compute backwards of functions, sometimes we don't need to compute the derivatives for everything. Just consider convolution between an input image and a learnable weight: while I can *mathematically* compute the derivative of the output with respect to the input and the weight, I don't actually need both of them; I only need the derivative for the learnable weight. `output_mask` is the parameter that tells me which gradients I need, and prevents me from doing unnecessary extra computation which will be thrown out in the end. The invariant is this: `output_mask[i]` is true IF AND ONLY IF `std::get<i>(result)` is defined.

Based on the way you've written the template, I think what you need to do is make `grad_input` defined or undefined depending on the setting of `output_mask`. That should give you the correct optional computation behavior, assuming you've written the template correctly (I don't know; I've only done a cursory review of the logic there.)

Reply:
Agreed. The issue is now fixed.